File: 3243-shortest-distance-after-road-addition-queries-i/shortest-distance-after-road-addition-queries-i.c
int* shortestDistanceAfterQueries(int n, int** queries, int queriesSize, int* queriesColSize, int* returnSize) {
    // Allocate graph adjacency list
    int** graph = (int**)malloc(n * sizeof(int*));
    int* graphSizes = (int*)calloc(n, sizeof(int));
    int* graphCapacities = (int*)malloc(n * sizeof(int));
    
    // Initial direct connections
    for (int i = 0; i < n; i++) {
        graph[i] = (int*)malloc(n * sizeof(int));
        graphCapacities[i] = n;
        
        if (i < n - 1) {
            graph[i][graphSizes[i]++] = i + 1;
        }
    }
    
    // Allocate result array
    int* answer = (int*)malloc(queriesSize * sizeof(int));
    *returnSize = queriesSize;
    
    // BFS helper function
    int bfs(int n, int** graph, int* graphSizes) {
        // Distance array
        int* dist = (int*)malloc(n * sizeof(int));
        for (int i = 0; i < n; i++) {
            dist[i] = INT_MAX;
        }
        
        // Queue for BFS
        int* queue = (int*)malloc(n * sizeof(int));
        int front = 0, rear = 0;
        
        // Start from city 0
        queue[rear++] = 0;
        dist[0] = 0;
        
        while (front < rear) {
            int current = queue[front++];
            
            // Reached destination
            if (current == n - 1) {
                int shortestPath = dist[n - 1];
                free(dist);
                free(queue);
                return shortestPath;
            }
            
            // Explore neighbors
            for (int i = 0; i < graphSizes[current]; i++) {
                int neighbor = graph[current][i];
                if (dist[neighbor] > dist[current] + 1) {
                    dist[neighbor] = dist[current] + 1;
                    queue[rear++] = neighbor;
                }
            }
        }
        
        free(dist);
        free(queue);
        return -1; // No path found
    }
    
    // Process each query
    for (int i = 0; i < queriesSize; i++) {
        int ui = queries[i][0];
        int vi = queries[i][1];
        
        // Add new road
        graph[ui][graphSizes[ui]++] = vi;
        
        // Find shortest path
        answer[i] = bfs(n, graph, graphSizes);
    }
    
    // Free memory
    for (int i = 0; i < n; i++) {
        free(graph[i]);
    }
    free(graph);
    free(graphSizes);
    free(graphCapacities);
    
    return answer;
}

File: 3243-shortest-distance-after-road-addition-queries-i/shortest-distance-after-road-addition-queries-i.java
class Solution {
    public int[] shortestDistanceAfterQueries(int n, int[][] queries) {
        // Initialize graph as an adjacency list
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        
        // Initial direct connections
        for (int i = 0; i < n - 1; i++) {
            graph.get(i).add(i + 1);
        }
        
        // Result array to store shortest path lengths
        int[] answer = new int[queries.length];
        
        // Process each query
        for (int i = 0; i < queries.length; i++) {
            // Add new road from ui to vi
            int ui = queries[i][0];
            int vi = queries[i][1];
            graph.get(ui).add(vi);
            
            // Find shortest path using BFS
            answer[i] = bfs(n, graph);
        }
        
        return answer;
    }
     private int bfs(int n, List<List<Integer>> graph) {
        // Queue for BFS traversal
        Queue<Integer> queue = new LinkedList<>();
        // Distance array to track shortest distances
        int[] dist = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        
        // Start from city 0
        queue.offer(0);
        dist[0] = 0;
        
        while (!queue.isEmpty()) {
            int current = queue.poll();
            
            // If reached destination
            if (current == n - 1) {
                return dist[n - 1];
            }
            
            // Explore neighbors
            for (int neighbor : graph.get(current)) {
                if (dist[neighbor] > dist[current] + 1) {
                    dist[neighbor] = dist[current] + 1;
                    queue.offer(neighbor);
                }
            }
        }
        
        return -1; // No path found
    }
}